### PR TITLE
Adding blog post on aggregations

### PR DESCRIPTION
**Pull request by gunnarmorling**

### CODE REVIEW

**Comment by gunnarmorling** [comment]
Hey @hpgrahsl, I'm coming back to your offer of reviewing this blog post :) So in case you have the time, I'd love to hear about any feedback you may have. But please don't feel stressed about it, it's perfectly fine if you can't do it.

File: blog/2018-09-20-consistent-aggregate-materialization-with-hibernate-and-debezium.adoc [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+Aggregates are serialized as JSON structure (which naturally can represent any nested object structure) and stored in a specific table.
+This is done as part of the actual transaction altering the data,
+which means the aggregate view is always consistent with the primary data.
+In particular this approach isn't prone to exposing intermediary aggregations as the KStreams-based solution discussed in the post linked above.
+
+The following picture shows the overall architecture:
+
+[.centered-image.responsive-image]
+++++
+<img src="/images/debezium-embedded.png" style="max-width:100%; margin-bottom:10px;" class="responsive-image" alt="Debezium Embedded Engine Streaming to Amazon Kinesis">
```
Comment (gunnarmorling) [inline blog/2018-09-20-consistent-aggregate-materialization-with-hibernate-and-debezium.adoc]: This still needs to be replaced with a correct picture.

**Comment by gunnarmorling** [comment]
Reworked a bit with wording updates etc.

File: blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+
+Following up to the link:/blog/2018/03/08/creating-ddd-aggregates-with-debezium-and-kafka-streams/[KStreams-based solution] to this we described recently,
+we'd like to present in this post an alternative for materializing such aggregate views driven by the application layer.
+
+== Overview
+
+The idea is to materialize views in a separate table in the source database,
+right in the moment the primary data is altered.
+
+Aggregates are serialized as JSON structure (which naturally can represent any nested object structure) and stored in a specific table.
```
Comment (hpgrahsl) [inline blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc]: plural? -> JSON structures

File: blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+This is done within the actual transaction altering the data,
+which means the aggregate view is always consistent with the primary data.
```
Comment (hpgrahsl) [inline blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc]: mixing / switching between the terms aggregates and aggregate views might be misleading?

File: blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+In particular this approach isn't prone to exposing intermediary aggregations as the KStreams-based solution discussed in the post linked above.
+
+The following picture shows the overall architecture:
+
+[.centered-image.responsive-image]
+++++
+<img src="/images/jpa_aggregations.png" style="max-width:100%; margin-bottom:10px;" class="responsive-image" alt="Streaming Materialized Aggregate Views to Elastisearch">
+++++
+
+Here the aggregate views are materialized by means of a small extension to http://hibernate.org/orm/[Hibernate ORM],
```
Comment (hpgrahsl) [inline blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc]: while conceptually comparable to a materialized view it's strictly speaking not such a thing we are dealing with here. at the end it's building aggregates which are written into a "normal table" or did I maybe misunderstand the approach taken?

File: blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+which stores the JSON aggregates within the source database.
+Changes to that aggregate table are then captured by Debezium and streamed to one topic per aggregate type.
+The Elasticsearch sink connector can subscribe to these topics and update corresponding full-text indexes.
+
+You can find a proof-of-concept implementation (said Hibernate extension and related code) of this idea in our https://github.com/debezium/debezium-examples/tree/master/end-to-end-demo[examples repository].
```
Comment (hpgrahsl) [inline blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc]: Is this the correct URL to the code?

File: blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+
+    ...
+}
+----
+
+Now if any entity annotated with `@MaterializeAggregate` is inserted, updated or deleted via Hibernate,
+the listener will kick in and materialize a JSON view of the aggregate root (customer) and its associated entities (addresses, category).
+
+Under the hood the https://github.com/FasterXML/jackson[Jackson API] is used for serializing the model into JSON.
+This means you can use any of its annotations to customize the JSON out, e.g. `@JsonIgnore`  to exclude the inverse relationship from `Address` to `Customer`:
```
Comment (hpgrahsl) [inline blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc]: JSON output

File: blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc [hunk 7]
```diff
@@ -0,0 +1,455 @@
...
+      "field" : "id"
+    } ],
+    "optional" : false,
+    "name" : "customers-complete.Key"
+  }
+}
+
+| { "id" : 1004 }
+
+| { "schema" : { ... } }
```
Comment (hpgrahsl) [inline blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc]: I assume the value schema isn't shown on purpose for brevity reasons?

File: blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc [hunk 8]
```diff
@@ -0,0 +1,455 @@
...
+This registers the MySQL connector with the "inventory" database
+(we're using an expanded version of the schema from the link:/docs/tutorial[Debezium tutorial]),
+capturing any changes to the "aggregates" table.
+
+==  Expanding JSON
+
+If we now were to browse the corresponding Kafka topic, we'd see data change events in the known Debezium format for all the changes to the `aggregates` table.
+
+The "materialization" field with the records' "after" state still is a single field containing a JSON string, though.
+What'd we'd rather like to have is a strongly typed Kafka Connect record, whose schema exactly describes the aggregate structure and the types of its fields.
```
Comment (hpgrahsl) [inline blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc]: typo? -> What we'd

File: blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc [hunk 9]
```diff
@@ -0,0 +1,455 @@
...
+          "total": 1
+      },
+      "timed_out": false,
+      "took": 11
+  }
+
+----
+
+And there you have it: a customer's complete data, including their addresses, categories, tags etc., materialized into a single document within Elasticsearch.
+If you're using JPA to update the customer, you'll see the data in index being updated accordingly in near-realtime.
```
Comment (hpgrahsl) [inline blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc]: ... in the index....

File: blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc [hunk 10]
```diff
@@ -0,0 +1,455 @@
...
+You also might argue that running JSON serializations within source transactions and storing aggregates within the source database represents some overhead.
+This often may be acceptable, though.
+
+Another question to ask is what's the advantage of using change data capture on an intermediary aggregate table over simply posting REST requests to Elasticsearch.
+The answer is the highly increased robustness and fault tolerance.
+If the Elasticsearch cluster can't be accessed for some reason, the machinery of Kafka and Kafka Connect will ensure that any change events will be propagated eventually, once the sink is up again.
+Also other consumers than Elasticsearch can subscribe to the aggregate topic, the log can be replayed from the beginning etc.
+
+Note that while we've been talking primarily about using Elasticsearch as a data sink, there are also other datastores and connectors that support complexly structured records.
+One example would be MongoDB and the https://github.com/hpgrahsl/kafka-connect-mongodb[sink connector] maintained by Hans-Peter Grahsl,
```
Comment (hpgrahsl) [inline blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc]: Blushing... thx for mentioning this :)

File: blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc [hunk 11]
```diff
@@ -0,0 +1,455 @@
...
+which one could use to sink customer aggregates into MongoDB, for instance enabling efficient retrieval of a customer and all their associated data with a single primary key look-up.
+
+== Outlook
+
+The Hibernate ORM extension as well as the SMT discussed in this post can be found in our https://github.com/debezium/debezium-examples/tree/master/end-to-end-demo[examples repository].
```
Comment (hpgrahsl) [inline blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc]: again is this the correct link? or maybe it's just not pushed yet :)

File: blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc [hunk 12]
```diff
@@ -0,0 +1,455 @@
...
+containing documents of customers with their addresses embedded,
+allowing you to efficiently search for customers based on their address.
+
+Following up to the link:/blog/2018/03/08/creating-ddd-aggregates-with-debezium-and-kafka-streams/[KStreams-based solution] to this we described recently,
+we'd like to present in this post an alternative for materializing such aggregate views driven by the application layer.
+
+== Overview
+
+The idea is to materialize views in a separate table in the source database,
+right in the moment the primary data is altered.
```
[the remark this answers is not in the document]
Comment (gunnarmorling) [reply blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc]: +1

File: blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc (hunk 4 above)
Comment (gunnarmorling) [reply blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc]: I agree I'm not 100% consistent with the terms and switch a bit forth and back. To me it definitely is a materialized view, though, just not in the sense of a SQL materialized view. I'd argue though that also in an RDBMS without native mat view support you could create materialized views yourself and store them in a table. And in a way that's what we're doing here.

File: blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc (hunk 5 above)
Comment (gunnarmorling) [reply blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc]: Good catch, fixed.

File: blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc (hunk 6 above)
Comment (gunnarmorling) [reply blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc]: Fixed.

File: blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc (hunk 8 above)
Comment (gunnarmorling) [reply blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc]: Fixed.

File: blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc (hunk 7 above)
Comment (gunnarmorling) [reply blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc]: Yes, right. I'll clarify that.

File: blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc (hunk 9 above)
Comment (gunnarmorling) [reply blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc]: Fixed.

File: blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc (hunk 11 above)
Comment (gunnarmorling) [reply blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc]: Thx, fixed it :)

File: blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc (hunk 10 above)
Comment (gunnarmorling) [reply blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc]: :) The least I can do.

File: blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc (hunk 3 above)
Comment (gunnarmorling) [reply blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc]: I double checked, it's "aggregate view" in most places. I think it's ok like that.

File: blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc (hunk 4 above)
Comment (gunnarmorling) [reply blog/2018-09-20-materializinga-ggregate-views-with-hibernate-and-debezium.adoc]: I've added a sentence to clarify things a bit.

**Comment by gunnarmorling** [comment]
Thanks for your review, @hpgrahsl! Addressed your remarks, great findings!

**Pull request merged**